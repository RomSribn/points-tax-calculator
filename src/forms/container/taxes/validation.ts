import * as yup from 'yup';

export const formSchema = yup.object().shape({
  income: yup
    .number()
    .typeError('You must specify a number')
    .min(1, 'The number must be greater than 0')
    .required('Income is required'),
  year: yup
    .number()
    .typeError('You must specify a number')
    .oneOf([2019, 2020, 2021, 2022], 'Only 2019, 2020, 2021, 2022 years are accessable')
    .required('Year is required')
});
