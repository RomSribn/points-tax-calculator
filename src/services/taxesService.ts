import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/dist/query/react';
import { ITaxesResponse } from '@utils/interfaces';
import { apiUrl } from '@utils/config';
/**
 * Main API service.
 */
const taxesService = createApi({
  reducerPath: 'taxesApi',
  baseQuery: fetchBaseQuery({ baseUrl: apiUrl }),
  tagTypes: ['Taxes'],
  endpoints: (build) => ({
    getTaxesBrackets: build.query<ITaxesResponse, number>({
      query: (year = 2022) => ({
        url: `/tax-calculator/tax-year/${year}`
      }),
      providesTags: () => ['Taxes']
    })
  })
});

export { taxesService };
