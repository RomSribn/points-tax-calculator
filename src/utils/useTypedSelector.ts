import { TypedUseSelectorHook, useSelector } from 'react-redux';
import { RootState } from '@redux/store';
/**
 * Typed useSelector.
 */
export const useTypedSelector: TypedUseSelectorHook<RootState> = useSelector;
