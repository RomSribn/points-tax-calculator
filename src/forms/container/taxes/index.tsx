import React, { useEffect } from 'react';
import { useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import { taxesService } from 'src/services/taxesService';
import { FormInput } from '@forms/controls';
import { setYear, setIncome, setResult } from '@redux/slices';
import { useDispatch } from 'react-redux';
import { useTypedSelector } from '@utils/useTypedSelector';
import { TYear } from '@utils/interfaces';
import { ErrorMessage } from '@forms/_elements/ErrorMessage';

import { formSchema } from './validation';

type TaxesFormFields = {
  income: number;
  year: TYear;
};

const TaxesForm: React.FC = () => {
  const {
    register,
    handleSubmit,
    formState: { errors }
  } = useForm<TaxesFormFields>({ resolver: yupResolver(formSchema) });
  const { year, income } = useTypedSelector((state) => state.app);
  const { data, isError, refetch, isFetching } = taxesService.useGetTaxesBracketsQuery(year);
  const dispatch = useDispatch();

  const onSubmit = handleSubmit(async (data) => {
    if (isError) {
      refetch();
      return;
    }

    dispatch(setYear(data.year));
    dispatch(setIncome(data.income));
  });

  useEffect(() => {
    if (!isFetching && income && data) {
      dispatch(setResult(data));
    }
  }, [data, dispatch, income, isFetching, year]);

  return (
    <div className="form-access my-auto" data-testid="taxes-form">
      <form onSubmit={onSubmit}>
        <span>Tax calculate</span>

        <FormInput<TaxesFormFields>
          id="income"
          type="text"
          name="income"
          label="income"
          placeholder="Type your income - $"
          register={register}
          errors={errors}
        />
        <FormInput<TaxesFormFields>
          id="year"
          type="text"
          name="year"
          label="year"
          placeholder="Type your tax year"
          register={register}
          errors={errors}
        />

        <button type="submit" className="btn btn-primary" disabled={isFetching}>
          {isFetching ? 'Calculating...' : 'Calculate'}
        </button>
        {isError && (
          <ErrorMessage className="mt-5 text-center text-danger">Something went wrong, please try again</ErrorMessage>
        )}
      </form>
    </div>
  );
};

export { TaxesForm };
