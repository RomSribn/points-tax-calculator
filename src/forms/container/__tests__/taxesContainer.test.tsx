import React from 'react';
import { customRender, screen, fireEvent } from '@utils/tests';
import { useTypedSelector } from '@utils/useTypedSelector';
import { taxesService } from '@services/taxesService';

import { TaxesForm } from '../taxes';

jest.mock('@services/taxesService');
jest.mock('@utils/useTypedSelector');
jest.mock('react-hook-form', () => ({
  ...jest.requireActual('react-hook-form'),
  useFormContext: () => ({
    handleSubmit: () => jest.fn(),
    getValues: () => jest.fn(),
    formState: {}
  }),
  useForm: () => ({ formState: {}, handleSubmit: jest.fn() })
}));

describe('TaxesForm', () => {
  // Mock the taxesService response
  const data = {};
  const apiState = {
    data,
    isError: false,
    refetch: jest.fn(),
    isFetching: false
  };
  beforeEach(() => {
    (useTypedSelector as jest.Mock).mockReturnValue({});
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Should renders the form', () => {
    (taxesService.useGetTaxesBracketsQuery as jest.Mock).mockReturnValue(apiState);

    const wrapper = customRender(<TaxesForm />);

    // Assert that the form elements are rendered
    expect(screen.getByLabelText('income')).toBeInTheDocument();
    expect(screen.getByLabelText('year')).toBeInTheDocument();
    expect(screen.getByText('Calculate')).toBeInTheDocument();

    expect(wrapper).toMatchSnapshot();
  });

  it('Should renders the form with loading button state', () => {
    (taxesService.useGetTaxesBracketsQuery as jest.Mock).mockReturnValue({ apiState, isFetching: true });

    const wrapper = customRender(<TaxesForm />);

    // Assert that the form elements are rendered
    expect(screen.getByText('Calculating...')).toBeInTheDocument();

    expect(wrapper).toMatchSnapshot();
  });

  it('Should displays an error message on submission error', async () => {
    // Mock the taxesService error response
    (taxesService.useGetTaxesBracketsQuery as jest.Mock).mockReturnValue({ ...apiState, data: null, isError: true });

    const wrapper = customRender(<TaxesForm />);

    // Submit the form
    fireEvent.submit(screen.getByText('Calculate'));

    // Assert that the error message is displayed
    const errorElement = await screen.findByText('Something went wrong, please try again');

    expect(errorElement).toBeInTheDocument();
    expect(wrapper).toMatchSnapshot();
  });
});
