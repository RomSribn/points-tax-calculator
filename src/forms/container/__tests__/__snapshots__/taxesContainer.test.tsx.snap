// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaxesForm Should displays an error message on submission error 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="form-access my-auto"
        data-testid="taxes-form"
      >
        <form>
          <span>
            Tax calculate
          </span>
          <div
            aria-live="polite"
            class=""
            data-testid="form-input"
          >
            <div
              class="form-group"
            >
              <input
                aria-invalid="false"
                aria-label="income"
                class="form-control"
                id="income"
                name="income"
                placeholder="Type your income - $"
                type="text"
              />
            </div>
          </div>
          <div
            aria-live="polite"
            class=""
            data-testid="form-input"
          >
            <div
              class="form-group"
            >
              <input
                aria-invalid="false"
                aria-label="year"
                class="form-control"
                id="year"
                name="year"
                placeholder="Type your tax year"
                type="text"
              />
            </div>
          </div>
          <button
            class="btn btn-primary"
            type="submit"
          >
            Calculate
          </button>
          <p
            class="font-serif text-sm text-left block text-red-600 mt-5 text-center text-danger"
            data-testid="error-message"
          >
            Something went wrong, please try again
          </p>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="form-access my-auto"
      data-testid="taxes-form"
    >
      <form>
        <span>
          Tax calculate
        </span>
        <div
          aria-live="polite"
          class=""
          data-testid="form-input"
        >
          <div
            class="form-group"
          >
            <input
              aria-invalid="false"
              aria-label="income"
              class="form-control"
              id="income"
              name="income"
              placeholder="Type your income - $"
              type="text"
            />
          </div>
        </div>
        <div
          aria-live="polite"
          class=""
          data-testid="form-input"
        >
          <div
            class="form-group"
          >
            <input
              aria-invalid="false"
              aria-label="year"
              class="form-control"
              id="year"
              name="year"
              placeholder="Type your tax year"
              type="text"
            />
          </div>
        </div>
        <button
          class="btn btn-primary"
          type="submit"
        >
          Calculate
        </button>
        <p
          class="font-serif text-sm text-left block text-red-600 mt-5 text-center text-danger"
          data-testid="error-message"
        >
          Something went wrong, please try again
        </p>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`TaxesForm Should renders the form 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="form-access my-auto"
        data-testid="taxes-form"
      >
        <form>
          <span>
            Tax calculate
          </span>
          <div
            aria-live="polite"
            class=""
            data-testid="form-input"
          >
            <div
              class="form-group"
            >
              <input
                aria-invalid="false"
                aria-label="income"
                class="form-control"
                id="income"
                name="income"
                placeholder="Type your income - $"
                type="text"
              />
            </div>
          </div>
          <div
            aria-live="polite"
            class=""
            data-testid="form-input"
          >
            <div
              class="form-group"
            >
              <input
                aria-invalid="false"
                aria-label="year"
                class="form-control"
                id="year"
                name="year"
                placeholder="Type your tax year"
                type="text"
              />
            </div>
          </div>
          <button
            class="btn btn-primary"
            type="submit"
          >
            Calculate
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="form-access my-auto"
      data-testid="taxes-form"
    >
      <form>
        <span>
          Tax calculate
        </span>
        <div
          aria-live="polite"
          class=""
          data-testid="form-input"
        >
          <div
            class="form-group"
          >
            <input
              aria-invalid="false"
              aria-label="income"
              class="form-control"
              id="income"
              name="income"
              placeholder="Type your income - $"
              type="text"
            />
          </div>
        </div>
        <div
          aria-live="polite"
          class=""
          data-testid="form-input"
        >
          <div
            class="form-group"
          >
            <input
              aria-invalid="false"
              aria-label="year"
              class="form-control"
              id="year"
              name="year"
              placeholder="Type your tax year"
              type="text"
            />
          </div>
        </div>
        <button
          class="btn btn-primary"
          type="submit"
        >
          Calculate
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`TaxesForm Should renders the form with loading button state 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="form-access my-auto"
        data-testid="taxes-form"
      >
        <form>
          <span>
            Tax calculate
          </span>
          <div
            aria-live="polite"
            class=""
            data-testid="form-input"
          >
            <div
              class="form-group"
            >
              <input
                aria-invalid="false"
                aria-label="income"
                class="form-control"
                id="income"
                name="income"
                placeholder="Type your income - $"
                type="text"
              />
            </div>
          </div>
          <div
            aria-live="polite"
            class=""
            data-testid="form-input"
          >
            <div
              class="form-group"
            >
              <input
                aria-invalid="false"
                aria-label="year"
                class="form-control"
                id="year"
                name="year"
                placeholder="Type your tax year"
                type="text"
              />
            </div>
          </div>
          <button
            class="btn btn-primary"
            disabled=""
            type="submit"
          >
            Calculating...
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="form-access my-auto"
      data-testid="taxes-form"
    >
      <form>
        <span>
          Tax calculate
        </span>
        <div
          aria-live="polite"
          class=""
          data-testid="form-input"
        >
          <div
            class="form-group"
          >
            <input
              aria-invalid="false"
              aria-label="income"
              class="form-control"
              id="income"
              name="income"
              placeholder="Type your income - $"
              type="text"
            />
          </div>
        </div>
        <div
          aria-live="polite"
          class=""
          data-testid="form-input"
        >
          <div
            class="form-group"
          >
            <input
              aria-invalid="false"
              aria-label="year"
              class="form-control"
              id="year"
              name="year"
              placeholder="Type your tax year"
              type="text"
            />
          </div>
        </div>
        <button
          class="btn btn-primary"
          disabled=""
          type="submit"
        >
          Calculating...
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
